internal/storageboxsubaccount/resource.go

```diff
@@ -0,0 +1,387 @@
+				},
+				"webdav_enabled": schema.BoolAttribute{
+					MarkdownDescription: "Whether the WebDAV subsystem is enabled.",
+					Optional:            true,
+					Computed:            true,
+					Default:             booldefault.StaticBool(false),
+				},
+				"readonly": schema.BoolAttribute{
+					MarkdownDescription: "Whether the Subaccount is read-only.",
+					Optional:            true,
+					Computed:            true,
+					Default:             booldefault.StaticBool(false),
+				},
+			},
+		},
+		"labels": resourceutil.LabelsSchema(),
+	}
+}
+
+type resourceModel struct {
+	model
+
+	Password types.String `tfsdk:"password"`
+}
+
+var _ util.ModelFromAPI[*hcloud.StorageBoxSubaccount] = &resourceModel{} // reuse model, as the fields from resourceModel are not readable anyway
+var _ util.ModelToTerraform[types.Object] = &resourceModel{}
+
+func (m *resourceModel) tfAttributesTypes() map[string]attr.Type {
+	return merge.Maps(
+		(&model{}).tfAttributesTypes(),
+		map[string]attr.Type{
+			"password": types.StringType,
+		},
+	)
+}
+
+func (m *resourceModel) ToTerraform(ctx context.Context) (types.Object, diag.Diagnostics) {
+	return types.ObjectValueFrom(ctx, m.tfAttributesTypes(), m)
+}
+
+func (r *Resource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var data resourceModel
+
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	storageBox := &hcloud.StorageBox{
+		ID: data.StorageBox.ValueInt64(),
+	}
+
+	opts := hcloud.StorageBoxSubaccountCreateOpts{
+		HomeDirectory: data.HomeDirectory.ValueString(),
+		Password:      data.Password.ValueString(),
+		Description:   data.Description.ValueString(),
+	}
+
+	resp.Diagnostics.Append(hcloudutil.TerraformLabelsToHCloud(ctx, data.Labels, &opts.Labels)...)
+
+	if !data.AccessSettings.IsUnknown() && !data.AccessSettings.IsNull() {
+		m := modelAccessSettings{}
+		resp.Diagnostics.Append(m.FromTerraform(ctx, data.AccessSettings)...)
+
+		opts.AccessSettings = &hcloud.StorageBoxSubaccountCreateOptsAccessSettings{
+			ReachableExternally: m.ReachableExternally.ValueBoolPointer(),
+			SambaEnabled:        m.SambaEnabled.ValueBoolPointer(),
+			SSHEnabled:          m.SSHEnabled.ValueBoolPointer(),
+			WebDAVEnabled:       m.WebDAVEnabled.ValueBoolPointer(),
+			Readonly:            m.Readonly.ValueBoolPointer(),
+		}
+	}
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Create in API
+	result, _, err := r.client.StorageBox.CreateSubaccount(ctx, storageBox, opts)
+	if err != nil {
+		resp.Diagnostics.Append(hcloudutil.APIErrorDiagnostics(err)...)
+		return
+	}
+
+	// Make sure to save the ID immediately so we can recover if the process stops after
+	// this call. Terraform marks the resource as "tainted", so it can be deleted and no
+	// surprise "duplicate resource" errors happen.
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("id"), types.Int64Value(result.Subaccount.ID))...)
```

Comment:
Should we save the parent storage box ID too? e.g. the user removes the resource from their config?

Reply:
Good catch, I assumed terraform would automagically add it to the state, as its a required attribute, but in a test the state only contained was `storage_box_id = null`, which caused the resource to be deleted from state as it could not be read, so it leaked.

Will add the attribute here and check the other resources for the same bug.